fvm/state/state.go

```diff
@@ -0,0 +1,162 @@
+package state
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+const (
+	defaultMaxKeySize         = 16_000     // ~16KB
+	defaultMaxValueSize       = 32_000_000 // ~32MB
+	defaultMaxInteractionSize = 64_000_000 // ~64MB
+)
+
+type StateOption func(st *State) *State
+type State struct {
+	ledger                Ledger
+	draft                 map[string]payload
+	interactionUsed       uint64
+	maxKeySizeAllowed     uint64
+	maxValueSizeAllowed   uint64
+	maxInteractionAllowed uint64
+}
+
+func defaultState(ledger Ledger) *State {
+	return &State{
+		ledger:                ledger,
+		interactionUsed:       uint64(0),
+		draft:                 make(map[string]payload, 0),
+		maxKeySizeAllowed:     defaultMaxKeySize,
+		maxValueSizeAllowed:   defaultMaxValueSize,
+		maxInteractionAllowed: defaultMaxInteractionSize,
+	}
+}
+
+// NewState constucts a new state
+func NewState(ledger Ledger, opts ...StateOption) *State {
+	ctx := defaultState(ledger)
+	for _, applyOption := range opts {
+		ctx = applyOption(ctx)
+	}
+	return ctx
+}
+
+// WithMaxKeySizeAllowed sets limit on max key size
+func WithMaxKeySizeAllowed(limit uint64) func(st *State) *State {
+	return func(st *State) *State {
+		st.maxKeySizeAllowed = limit
+		return st
+	}
+}
+
+// WithMaxValueSizeAllowed sets limit on max value size
+func WithMaxValueSizeAllowed(limit uint64) func(st *State) *State {
+	return func(st *State) *State {
+		st.maxValueSizeAllowed = limit
+		return st
+	}
+}
+
+// WithMaxInteractionAllowed sets limit on total byte interaction with ledger
+func WithMaxInteractionAllowed(limit uint64) func(st *State) *State {
+	return func(st *State) *State {
+		st.maxInteractionAllowed = limit
+		return st
+	}
+}
+
+func (s State) Read(owner, controller, key string) (flow.RegisterValue, error) {
+	if err := s.checkSize(owner, controller, key, []byte{}); err != nil {
+		return nil, err
+	}
+
+	// check draft first
+	if p, ok := s.draft[fullKey(owner, controller, key)]; ok {
+		err := s.ledger.Touch(owner, controller, key)
+		if err != nil {
+			return nil, &LedgerFailure{err}
+		}
+		return p.value, nil
+	}
+
+	value, err := s.ledger.Get(owner, controller, key)
+	if err != nil {
+		return nil, &LedgerFailure{err}
+	}
+	// update interaction
+	keySize := uint64(len(owner) + len(controller) + len(key))
+	valueSize := uint64(len(value))
+	err = s.updateInteraction(keySize, valueSize, 0)
+	if err != nil {
+		return nil, &LedgerFailure{err}
+	}
+
+	return value, nil
+}
+
+func (s State) Update(owner, controller, key string, value flow.RegisterValue) error {
```

Comment:
I'm wondering if it would simplify things to use `flow.RegisterID` as a composite parameter here. It was trickier on the `Ledger` interface, but I don't see why not to accept it here

Reply:
I have a plan for flow.Registers but won't do it in this PR to keep the scope of file changes small.